.github/scripts/filter_changed_files.sh

```diff
@@ -0,0 +1,51 @@
+#!/bin/bash
+
+set -euo pipefail
+
+trigger_ci=false
+
+# If we are on main we want to always run the tests
+if [[ $GITHUB_REF_NAME == "main" ]]; then
+	trigger_ci=true
+	echo "On main, running CI: $trigger_ci"
+	echo "trigger-ci=$trigger_ci" >>"$GITHUB_OUTPUT"
+	exit 0
+fi
+
+# ensure something was passed in (this is not meant to be that robust)
+if [ $# -eq 0 ]; then
+	echo "No parameters, triggering CI: $trigger_ci"
+	echo "trigger-ci=$trigger_ci" >>"$GITHUB_OUTPUT"
+	exit 0
+fi
+
+# Using `git merge-base` ensures that we're always comparing against the correct branch point. 
+#For example, given the commits:
+#
+# A---B---C---D---W---X---Y---Z # origin/main
+#             \---E---F         # feature/branch
+#
+# ... `git merge-base origin/main` would return commit `D`
+# `...HEAD` specifies from the common ancestor to the latest commit on the current branch (HEAD)..
+files_to_check=$(git diff --name-only "$(git merge-base origin/main HEAD~)"...HEAD)
```

Comment:
shamelessly lifted this from [consul's implementation](https://github.com/hashicorp/consul/blob/main/.github/scripts/filter_changed_files_go_test.sh#L16) as I'm 100% sure they know what they're doing compared to me here